cbmc/proofs/IotMqtt_Notify/_IotMqtt_Notify_harness.c

```diff
@@ -0,0 +1,115 @@
+                                    void * pCompare );
+
+typedef void ( *FreeElementFunction_t )( void * pData );
+
+/**
+ * We constrain the return values of these functions because
+ * they are checked by assertions in the MQTT code.
+ */
+IotTaskPoolError_t IotTaskPool_CreateJob( IotTaskPoolRoutine_t userCallback,
+                                          void * pUserContext,
+                                          IotTaskPoolJobStorage_t * const pJobStorage,
+                                          IotTaskPoolJob_t * const pJob )
+{
+    assert( userCallback != NULL );
+    assert( pJobStorage != NULL );
+    assert( pJob != NULL );
+
+    /* _IotMqtt_ScheduleOperation asserts this. */
+    return IOT_TASKPOOL_SUCCESS;
+}
+
+IotTaskPoolError_t IotTaskPool_ScheduleDeferred( IotTaskPool_t taskPool,
+                                                 IotTaskPoolJob_t job,
+                                                 uint32_t timeMs )
+{
+    IotTaskPoolError_t error;
+
+    /* _IotMqtt_ScheduleOperation asserts this. */
+    __CPROVER_assume( error != IOT_TASKPOOL_BAD_PARAMETER &&
+                      error != IOT_TASKPOOL_ILLEGAL_OPERATION );
+    return error;
+}
+
+/**
+ * We assume the list remove functions are memory safe.
+ *
+ * We abstract the list remove functions for performance reasons.  Our
+ * abstraction replaces the original list with an unconstrained list.
+ * Our abstraction proves that none of the elements on the original
+ * list are accessed after the remove: we free all elements on the
+ * original list, so that any later access will be caught as a
+ * use-after-free error.
+ */
+void IotListDouble_RemoveAllMatches( const IotListDouble_t * const pList,
+                                     MatchFunction_t isMatch,
+                                     void * pMatch,
+                                     FreeElementFunction_t freeElement,
+                                     size_t linkOffset )
+{
+    free_IotMqttSubscriptionList( pList );
+    allocate_IotMqttSubscriptionList( pList, SUBSCRIPTION_COUNT_MAX - 1 );
+}
+
+void harness()
+{
+    /* Assume a valid MQTT connection. */
+    IotMqttConnection_t pMqttConnection = allocate_IotMqttConnection( NULL );
+    __CPROVER_assume( pMqttConnection != NULL );
+    __CPROVER_assume( pMqttConnection->pNetworkInterface != NULL );
+    __CPROVER_assume( IS_STUBBED_NETWORKIF_SEND( pMqttConnection->pNetworkInterface ) );
+    __CPROVER_assume( IS_STUBBED_NETWORKIF_DESTROY( pMqttConnection->pNetworkInterface ) );
+    ensure_IotMqttConnection_has_lists( pMqttConnection );
+    __CPROVER_assume( valid_IotMqttConnection( pMqttConnection ) );
+    __CPROVER_assume( pMqttConnection->references > 0 );
+
+    /* Assume unconstrained operation. */
+    IotMqttOperation_t pOperation = allocate_IotMqttOperation( NULL, pMqttConnection );
+    __CPROVER_assume( valid_IotMqttOperation( pOperation ) );
+    __CPROVER_assume( pOperation->incomingPublish == false ); 
```

Comment:
Is this true?  Give a comment saying why.

Reply:
Yes, incoming publish messages are processed by ProcessIncomingPublish (I added a comment).

Reply:
I think a better comment would be that "Inbound packets are either an inbound publish or an inbound response (a ping response or an acknowledgement).  The purpose of Notify is to alert any task waiting on an inbound response. Notify is never invoked on an inbound publish."

Reply:
Implemented as suggested.

Reply:
Good.